avocado/core/test.py

```diff
@@ -133,7 +133,136 @@ def str_filesystem(self):
                                  % (self.str_uid, str(self)))
 
 
-class Test(unittest.TestCase):
+class BaseTestData(object):
+
+    """
+    Class that adds the hability for tests to have access to data files
+
+    An implementation of this interfaces needs to declare the name of
+    data sources it knows, and a single method for users to get the
+    data files, either from a specific source, or from its built-in
+    order of sources.
+    """
+
+    DATA_SOURCES = []
+
+    def _check_valid_data_source(self, source):
+        """
+        Utility to check if user chose a specific data source
+
+        :param source: either None for no specific selection or a source name
+        :type source: None or str
+        :raises: ValueError
+        """
+        if source is not None and source not in self.DATA_SOURCES:
+            msg = 'Data file source requested (%s) is not one of: %s'
+            msg %= (source, ', '.join(self.DATA_SOURCES))
+            raise ValueError(msg)
+
+    def get_data(self, filename, source=None, must_exist=True):
+        """
+        Retrieves the path to a given data file.
+
+        :param filename: the name of the data file to be retrieved
+        :type filename: str
+        :param source: one of the defined data sources
+        :type source: str
+        :param must_exist: whether the existence of a file is checked
+        :type must_exist: bool
+        :rtype: str or None
+        """
+        raise NotImplementedError
+
+
+class TestData(BaseTestData):
+
+    DATA_SOURCES = ["file", "test", "variant"]
+
+    def __init__(self):
+        self._data_sources_mapping = {
+            "file": [lambda: self.datadir],
+            "test": [lambda: self.datadir,
+                     lambda: "%s.%s" % (self.__class__.__name__,
+                                        self._testMethodName)],
+            "variant": [lambda: self.datadir,
+                        lambda: "%s.%s" % (self.__class__.__name__,
+                                           self._testMethodName),
+                        lambda: self.name.variant]
+        }
+
+    def _get_datadir(self, source):
+        path_components = self._data_sources_mapping.get(source)
+        if path_components is None:
+            return
+
+        # evaluate lazily, needed when the class changes its own
+        # information such as its datadir
```

Comment:
The lambdas was the spark (sorry again). Anyway reading this commit I understand the usage of it. On the other hand this class should be initialized after `Test` init, so the datadir should be probably ready. What troubles me is that with this way it'd be possible to get 2 different datadirs throughout a test execution, which would be IMO wrong. Or is it what you are trying to allow?

Anyway after reading the sources again today morning I don't have anything against this interface, only some tweaks regarding the functionality, but that can be tweaked later (interface can't that easily, that was the point of all the versions).